### PR TITLE
support query param regex matching in Gateway API

### LINKED ISSUE #5262: Gateway API: support query param regex matching
As a follow-up to https://github.com/projectcontour/contour/pull/5239, [query parameter matching also allows for RegularExpression matches as custom conformance](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.QueryParamMatchType). It would be great to add support for this.

### PR DESCRIPTION
**Pull request by padlar**
fixes #5262

### CODE REVIEW

**Comment by skriss** [comment]
Thanks @padlar! Will review soon.

File: internal/dag/gatewayapi_processor.go [hunk 1]
```diff
@@ -1667,10 +1667,17 @@ func gatewayQueryParamMatchConditions(matches []gatewayapi_v1beta1.HTTPQueryPara
...
-		// the only supported match type.
-		if match.Type != nil && *match.Type != gatewayapi_v1beta1.QueryParamMatchExact {
-			return nil, fmt.Errorf("HTTPRoute.Spec.Rules.Matches.QueryParams: Only Exact match type is supported")
+		var matchType string
+		switch ref.Val(match.Type, gatewayapi_v1beta1.QueryParamMatchExact) {
+		case gatewayapi_v1beta1.QueryParamMatchExact:
+			matchType = HeaderMatchTypeExact
+		case gatewayapi_v1beta1.QueryParamMatchRegularExpression:
+			if err := ValidateRegex(match.Value); err != nil {
+				return nil, fmt.Errorf("FIXME: Invalid value for RegularExpression match type is specified")
```
Comment (skriss) [inline internal/dag/gatewayapi_processor.go:1676]:
```suggestion
				return nil, fmt.Errorf("HTTPRoute.Spec.Rules.Matches.QueryParams: Invalid value for RegularExpression match type is specified")
```

File: internal/dag/gatewayapi_processor.go [hunk 2]
```diff
@@ -1667,10 +1667,17 @@ func gatewayQueryParamMatchConditions(matches []gatewayapi_v1beta1.HTTPQueryPara
...
+			}
+			matchType = HeaderMatchTypeRegex
+		default:
+			return nil, fmt.Errorf("FIXME: Only support Exact or RegularExpression match types")
```
Comment (skriss) [inline internal/dag/gatewayapi_processor.go:1680]:
```suggestion
			return nil, fmt.Errorf("HTTPRoute.Spec.Rules.Matches.QueryParams: Only Exact and RegularExpression match types are supported")
```

Comment (skriss) [inline internal/dag/gatewayapi_processor.go:1680]: Let's add a test case to status_test.go that covers this branch (unsupported query param match type)

**Review by skriss — CHANGES_REQUESTED** [review]
Thanks @padlar! This is looking pretty good, just a few comments. Please also add some test cases to internal/dag/builder_test.go in the `TestDAGInsertGatewayAPI` test -- you should be able to find some existing HTTPRoute query param tests to base the new test cases off of.

File: internal/dag/gatewayapi_processor.go (hunk 2 above)
Comment (padlar) [reply internal/dag/gatewayapi_processor.go]: Added few more test cases

**Comment by skriss** [comment]
Thanks for the updates @padlar. You'll also need to add a changelog file, per https://github.com/projectcontour/contour/actions/runs/4824508057/jobs/8594362044?pr=5310 and https://github.com/projectcontour/contour/blob/main/CONTRIBUTING.md#commit-message-and-pr-guidelines, so users can read about this in the release notes!

**Review by skriss — APPROVED** [review]
LGTM, thanks @padlar!

**Pull request merged**